### PR TITLE
Show chapter highlighted through playback

### PR DESCRIPTION
**Pull request by mr-mr**
Chapters on top of skipping, also highlight when within the chapter in the video

image

image

### CODE REVIEW

File: src/css/_docs.scss [hunk 1]
```diff
@@ -418,12 +430,16 @@
...
+      // background-color: darken($c-background, 3%);
     }
     &:before {
       color: $c-white;
       counter-increment: chapters;
       content: "Chapter " counter(chapters) ": ";
     }
   }
+  &-active {
+    border: 1px solid $c-white !important;
```
Comment (akash1810) [inline src/css/_docs.scss]: `!important` is always a warning sign. If its the only thing that works, can we add a comment to explain why?

File: src/js/lib/youtube.js [hunk 2]
```diff
@@ -13,7 +14,37 @@ export function pimpYouTubePlayer(videoId, node, height, width) {
                 events: {
                     'onReady': function(){
                       resolve(youTubePlayer);
-                    }
+                      // checkPlaybackTime(youTubePlayer);
```
Comment (akash1810) [inline src/js/lib/youtube.js]: commented code doesn't get run, can we remove this?

File: src/js/main.js [hunk 3]
```diff
@@ -18,6 +18,12 @@ export function init(el, context, config) {
         const youTubeId = resp.sheets[config.sheetName][0].youTubeId;
         const chapters = resp.sheets[config.sheetChapter];
 
+        chapters.forEach( function(chapter, index){
+          if(chapters.length > index+1){
+            var endChapter = chapters[index+1];
+            chapter.endChapter = endChapter.chapterTimestamp -1;
```
Comment (akash1810) [inline src/js/main.js]: I think `endChapter.chapterTimestamp` is a string at this point, so we'd need to `parseInt(endChapter.chapterTimestamp)` before `-1` for safety

File: src/js/lib/youtube.js [hunk 4]
```diff
@@ -13,7 +14,37 @@ export function pimpYouTubePlayer(videoId, node, height, width) {
...
+                            if(playerCurrentTime >= chapStart && playerCurrentTime <= chapEnd){
+                              return value;
+                            }
+                          });
+                          if(currentChapter.length === 1){
+                            var chapterAll = [].slice.call(document.querySelectorAll('li[data-sheet-timestamp]'));
+                            chapterAll.forEach( function(el){
+                              if(el.dataset.sheetTimestamp === currentChapter[0].chapterTimestamp){
+                                el.classList.add('docs--chapters-active');
+                              }else{
```
Comment (akash1810) [inline src/js/lib/youtube.js]: very minor: the spacing here is making me 😢...

File: src/js/lib/youtube.js [hunk 5]
```diff
@@ -13,7 +14,37 @@ export function pimpYouTubePlayer(videoId, node, height, width) {
...
                       resolve(youTubePlayer);
-                    }
+                      // checkPlaybackTime(youTubePlayer);
+                    },
+                    'onStateChange': function(event){
+                      let chapTimer;
+                      if (event.data == YT.PlayerState.PLAYING) {
+
+                        var playerTotalTime = youTubePlayer.getDuration();
+                        chapTimer = setInterval(function() {
```
Comment (akash1810) [inline src/js/lib/youtube.js]: should we test performance of this on mobile too; running this every second may be too much for mobile?

File: src/js/config.js [hunk 6]
```diff
@@ -24,6 +24,7 @@ System.config({
     "core-js": "npm:core-js@1.2.6",
     "guardian/iframe-messenger": "github:guardian/iframe-messenger@master",
     "json": "github:systemjs/plugin-json@0.1.2",
+    "moment": "npm:moment@2.14.1",
```
Comment (akash1810) [inline src/js/config.js]: we don't seem to be using momentjs (yet) so do we need it?

**Comment by akash1810** [comment]
mostly minor comments, but LGTM 👍

File: src/js/main.js [hunk 7]
```diff
@@ -18,6 +18,12 @@ export function init(el, context, config) {
         const youTubeId = resp.sheets[config.sheetName][0].youTubeId;
         const chapters = resp.sheets[config.sheetChapter];
 
+        chapters.forEach( function(chapter, index){
```
Comment (akash1810) [inline src/js/main.js]: might be worth sorting chapters by `chapterTimestamp` ascending too (never trust external data 😜)

File: src/js/main.js [hunk 8]
```diff
@@ -18,6 +18,15 @@ export function init(el, context, config) {
         const youTubeId = resp.sheets[config.sheetName][0].youTubeId;
         const chapters = resp.sheets[config.sheetChapter];
 
+        chapters.sort(chapters.chapterTimestamp);
+
+        chapters.forEach( function(chapter, index){
+          if(chapters.length > index+1){
+            var endChapter = chapters[index+1];
+            parseInt(endChapter.chapterTimestamp);
+            chapter.endChapter = endChapter.chapterTimestamp -1;
```
Comment (akash1810) [inline src/js/main.js]: the `parseInt` is in the wrong place; I think you'd want

```js
chapter.endChapter = parseInt(endChapter.chapterTimestamp) - 1;
```

File: src/js/main.js [hunk 9]
```diff
@@ -18,19 +18,27 @@ export function init(el, context, config) {
         const youTubeId = resp.sheets[config.sheetName][0].youTubeId;
         const chapters = resp.sheets[config.sheetChapter];
 
+        chapters.sort((a,b) => parseInt(a.chapterTimestamp) - parseInt(b.chapterTimestamp));
+
+        chapters.forEach( function(chapter, index){
+          if(chapters.length > index+1){
+            const endChapter = chapters[index+1];
```
Comment (akash1810) [inline src/js/main.js]: would `nextChapter` be a more explicit variable name?

**Comment by akash1810** [comment]
`const` all the things!!

Looks good 👍

File: src/js/main.js [hunk 10]
```diff
@@ -18,19 +18,27 @@ export function init(el, context, config) {
         const youTubeId = resp.sheets[config.sheetName][0].youTubeId;
         const chapters = resp.sheets[config.sheetChapter];
 
+        chapters.sort((a,b) => parseInt(a.chapterTimestamp) - parseInt(b.chapterTimestamp));
+
+        chapters.forEach( function(chapter, index){
```
Comment (gidsg) [inline src/js/main.js]: very tiny, but you don't need the space here before `function`

**Comment by gidsg** [comment]
Looks good 🚀 👍

**Pull request merged**